include/vcpkg/base/expected.h

```diff
@@ -121,101 +41,172 @@ namespace vcpkg
         ExpectedHolder() : t(nullptr) { }
         using pointer = T*;
         using const_pointer = T*;
-        T* get() { return t; }
-        T* get() const { return t; }
+        T* get() noexcept { return t; }
+        T* get() const noexcept { return t; }
         T* t;
     };
 
     template<class T, class S>
     struct ExpectedT
     {
-        constexpr ExpectedT() = default;
-
         // Constructors are intentionally implicit
-
-        ExpectedT(const S& s, ExpectedRightTag = {}) : m_s(s) { }
+        ExpectedT(const S& s, ExpectedRightTag = {}) : m_s(s), value_is_error(true) { }
         template<class U = S, std::enable_if_t<!std::is_reference<U>::value, int> = 0>
-        ExpectedT(S&& s, ExpectedRightTag = {}) : m_s(std::move(s))
+        ExpectedT(S&& s, ExpectedRightTag = {}) : m_s(std::move(s)), value_is_error(true)
         {
         }
 
-        ExpectedT(const T& t, ExpectedLeftTag = {}) : m_t(t) { }
+        ExpectedT(const T& t, ExpectedLeftTag = {}) : m_t(t), value_is_error(false) { }
         template<class U = T, std::enable_if_t<!std::is_reference<U>::value, int> = 0>
-        ExpectedT(T&& t, ExpectedLeftTag = {}) : m_t(std::move(t))
+        ExpectedT(T&& t, ExpectedLeftTag = {}) : m_t(std::move(t)), value_is_error(false)
         {
         }
 
-        ExpectedT(const ExpectedT&) = default;
-        ExpectedT(ExpectedT&&) = default;
-        ExpectedT& operator=(const ExpectedT&) = default;
-        ExpectedT& operator=(ExpectedT&&) = default;
+        ExpectedT(const ExpectedT& other) : value_is_error(other.value_is_error)
+        {
+            if (value_is_error)
+            {
+                ::new (&m_s) S(other.m_s);
+            }
+            else
+            {
+                ::new (&m_t) ExpectedHolder<T>(other.m_t);
+            }
+        }
 
-        explicit constexpr operator bool() const noexcept { return !m_s.has_error(); }
-        constexpr bool has_value() const noexcept { return !m_s.has_error(); }
+        ExpectedT(ExpectedT&& other) : value_is_error(other.value_is_error)
+        {
+            if (value_is_error)
+            {
+                ::new (&m_s) S(std::move(other.m_s));
+            }
+            else
+            {
+                ::new (&m_t) ExpectedHolder<T>(std::move(other.m_t));
+            }
+        }
+
+        // copy assign is deleted to avoid creating "valueless by exception" states
+        ExpectedT& operator=(const ExpectedT& other) = delete;
+
+        ExpectedT& operator=(ExpectedT&& other) noexcept // enforces termination
+        {
+            if (value_is_error)
+            {
+                if (other.value_is_error)
+                {
+                    m_s = std::move(other.m_s);
+                }
+                else
+                {
+                    m_s.~S();
```

Comment:
guard the explicit destructor call with `if constexpr (std::is_class_v<S>) ` ?

Reply:
I don't believe we can use if constexpr unconditionally with our current compiler support matrix. (And also it's not really meaningful unless you could make the whole Expected into a trivial type which is almost never a thing.

Reply:
> I don't believe we can use if constexpr unconditionally with our current compiler support matrix. 

Isn't vcpkg requiring c++17?

> unless you could make the whole Expected into a trivial type which is almost never a thing.

another assumption another `static_assert` ?

Reply:
> > I don't believe we can use if constexpr unconditionally with our current compiler support matrix.
> 
> Isn't vcpkg requiring c++17?

Requiring a specific standard version is almost valueless: what matters is what the implementations one targets support, and many of those get individual features long before they get a whole standard.

> > unless you could make the whole Expected into a trivial type which is almost never a thing.
> 
> another assumption another `static_assert` ?

I'm not sure I understand. The answer to "is this trivial" will almost always be "no", so why static_assert?

Reply:
Ok i thought the following wouldn't compile
```
int myint = 255;
myint.~int();
```
but apparently it does and does nothing.